c++/src/capnp/CMakeLists.txt

```diff
@@ -269,17 +269,26 @@ if(BUILD_TESTING)
     serialize-packed-test.c++
     canonicalize-test.c++
     fuzz-test.c++
-    test-util.c++
-    ${test_capnp_cpp_files}
-    ${test_capnp_h_files}
   )
-  target_link_libraries(capnp-tests ${test_libraries})
-  add_dependencies(capnp-tests test_capnp)
-  add_dependencies(check capnp-tests)
-  add_test(NAME capnp-tests-run COMMAND capnp-tests)
+  foreach(TEST_FILE IN LISTS TEST_FILES)
+    get_filename_component(exe_dir "${TEST_FILE}" DIRECTORY)
+    get_filename_component(exe_name "${TEST_FILE}" NAME_WE)
+    if(exe_dir)
+      set(target "capnp-${exe_dir}-${exe_name}")
+      set(exe_name "${exe_dir}/${exe_name}")
+    else()
+      set(target capnp-${exe_name})
+    endif()
+    add_executable(${target} ${TEST_FILE} test-util.c++ ${test_capnp_cpp_files} ${test_capnp_h_files})
+    set_target_properties(${target} PROPERTIES OUTPUT_NAME ${exe_name})
+    target_link_libraries(${target} ${test_libraries})
+    add_dependencies(${target} test_capnp)
+    add_dependencies(check ${target})
+    add_test(NAME ${target}-run COMMAND ${exe_name}${CMAKE_EXECUTABLE_SUFFIX})
```

Comment:
To avoid code duplication, I think you could factor out the guts of this loop into a CMake function, maybe named `capnproto_add_test()` and stick it in the top-level CMakeLists.txt (`capnproto/c++/CMakeLists.txt`). Alternatively, it could live in a new file maybe called `capnproto/c++/cmake/CapnProtoAddTest.cmake`, and you could `include(CapnProtoAddTest)` where it's needed.